src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateConverter.cs

```diff
@@ -14,8 +14,62 @@
 
 namespace GitHub.DistributedTask.Pipelines.ObjectTemplating
 {
-    internal static class PipelineTemplateConverter
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class PipelineTemplateConverter
     {
+        public static List<Step> ConvertToSteps(
+            TemplateContext context,
+            TemplateToken steps)
+        {
+            var stepsSequence = steps.AssertSequence($"job {PipelineTemplateConstants.Steps}");
+
+            var result = new List<Step>();
+            var nameBuilder = new ReferenceNameBuilder();
+            foreach (var stepsItem in stepsSequence)
+            {
+                var step = ConvertToStep(context, stepsItem, nameBuilder);
+                if (step != null) // step = null means we are hitting error during step conversion, there should be an error in context.errors
+                {
+                    if (step.Enabled)
+                    {
+                        result.Add(step);
+                    }
+                }
+            }
+
+            // Generate context name if empty
+            foreach (ActionStep step in result)
+            {
+                if (!String.IsNullOrEmpty(step.ContextName))
+                {
+                    continue;
+                }
+
+                var name = default(string);
+                switch (step.Reference.Type)
+                {
+                    case ActionSourceType.ContainerRegistry:
+                        var containerReference = step.Reference as ContainerRegistryReference;
+                        name = containerReference.Image;
```

Comment:
what will `namebuilder` do for special characters in docker image, like `:`?

Reply:
replace with _